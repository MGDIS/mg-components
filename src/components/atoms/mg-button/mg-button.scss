/* Variables */
$mg-button-border-width: 0.1rem;
$mg-button-padding-horizontal: 1.1rem;
$button-column-gap: 0.6rem;

/* Mixins */
@mixin style-gradient($name, $backgroundColor, $fontColor, $hasVariable: false) {
  @if $hasVariable {
    $backgroundColor: --mg-button-#{$name}-color;
    $fontColor: --mg-button-#{$name}-font-color;
  } @else {
    $backgroundColor: --color-#{$backgroundColor};
    $fontColor: --color-#{$fontColor};
  }
  .mg-button.mg-button--#{$name} {
    // Decoration
    border-color: hsl(
      var(
        #{--mg-button-#{$name}-border-color},
        (var(#{$backgroundColor}-h), var(#{$backgroundColor}-s), calc(var(#{$backgroundColor}-l) - var(--mg-button-border-variation) * 10%))
      )
    );
    // Font
    color: hsl(var(#{$fontColor}));
    // Background
    background: linear-gradient(
      to bottom,
      hsl(var(#{$backgroundColor}-h), var(#{$backgroundColor}-s), var(#{$backgroundColor}-l)),
      hsl(var(#{$backgroundColor}-h), var(#{$backgroundColor}-s), calc(var(#{$backgroundColor}-l) - var(--mg-button-gradient) * 10%))
    );
  }
  // Only apply if not disabled
  :host(:hover),
  :host(:focus),
  :host(:active) {
    .mg-button--#{$name}:not(.mg-button--disabled) {
      background: hsl(var(#{$backgroundColor}-h), var(#{$backgroundColor}-s), var(#{$backgroundColor}-l));
    }
  }
}

@mixin font-style() {
  text-decoration: none;
  font-family: inherit;
  font-weight: var(--mg-button-font-weight, normal);
  text-align: left;
  font-size: var(--font-size);
}

:host {
  display: inline-block;
  vertical-align: middle;
  max-width: 100%;
}

:host(:focus-visible) {
  outline: 0;
  // same style as chrome
  .mg-button:not(.mg-button--disabled) {
    outline-color: -webkit-focus-ring-color;
    outline-style: auto;
    outline-width: 0.2rem;
  }
}

// Only apply if not disabled
:host(:hover),
:host(:focus),
:host(:active) {
  .mg-button--flat:not(.mg-button--disabled) {
    background: hsla(var(--color-dark), 0.1);
  }
}

:host([full-width]) {
  display: block;
}

// Gradient style
@include style-gradient('primary', null, null, true);
@include style-gradient('secondary', null, null, true);
@include style-gradient('danger', 'danger', 'neutral');
@include style-gradient('danger-alt', 'neutral', 'danger');
@include style-gradient('info', 'info', 'neutral');
@include style-gradient('success', 'success', 'dark');

/* Styles */
.mg-button {
  // Display
  position: relative;
  display: inline-flex;
  align-items: center;
  column-gap: $button-column-gap;
  padding: calc((var(--default-size) - (2 * $mg-button-border-width) - (var(--font-size) * var(--line-height))) / 2) $mg-button-padding-horizontal;
  min-height: var(--default-size);
  box-sizing: border-box;
  // Decoration
  border-top-right-radius: var(--mg-button-border-radius-top-right, var(--mg-button-border-radius));
  border-top-left-radius: var(--mg-button-border-radius-top-left, var(--mg-button-border-radius));
  border-bottom-right-radius: var(--mg-button-border-radius-bottom-right, var(--mg-button-border-radius));
  border-bottom-left-radius: var(--mg-button-border-radius-bottom-left, var(--mg-button-border-radius));
  cursor: pointer;
  background-image: none;
  border-top-width: var(--mg-button-border-top-width, $mg-button-border-width);
  border-right-width: var(--mg-button-border-right-width, $mg-button-border-width);
  border-bottom-width: var(--mg-button-border-bottom-width, $mg-button-border-width);
  border-left-width: var(--mg-button-border-left-width, $mg-button-border-width);
  border-style: solid;
  // Font (necessary for stylesheet only)
  @include font-style();
  // Other
  touch-action: manipulation;

  // Disabled style
  &.mg-button--disabled {
    opacity: var(--mg-button-disabled-opacity);
    cursor: default;
  }

  // Icon only button
  &.mg-button--icon {
    // Display
    padding: 0;
    justify-content: center;
    width: var(--default-size);
    height: var(--default-size);
    // Decoration
    border-top-right-radius: var(--mg-button-border-radius-top-right, var(--mg-button-icon-border-radius));
    border-top-left-radius: var(--mg-button-border-radius-top-left, var(--mg-button-icon-border-radius));
    border-bottom-right-radius: var(--mg-button-border-radius-bottom-right, var(--mg-button-icon-border-radius));
    border-bottom-left-radius: var(--mg-button-border-radius-bottom-left, var(--mg-button-icon-border-radius));
    // font
    font-size: var(--font-size);
    line-height: var(--font-size);
  }

  // Flat style
  &.mg-button--flat {
    background: none;
    border: none;
  }
  // full-width style
  &.mg-button--full-width {
    display: flex;
    justify-content: center;
  }

  // loading style
  &.mg-button--loading {
    > mg-icon:first-of-type {
      position: absolute;
      left: 50%;
      top: 50%;
      transform: translate(-50%, -50%);
    }
    .mg-button__content {
      opacity: 0;
    }
  }

  mg-icon:only-child,
  ::slotted(mg-icon) {
    display: inline-flex;
    flex-shrink: 0;
    line-height: 1;
    margin: auto 0;
  }

  slot,
  ::slotted(*:not(mg-icon)) {
    // Font
    @include font-style();
    // Display
    display: flex;
    align-items: center;
    column-gap: $button-column-gap;
  }
}
