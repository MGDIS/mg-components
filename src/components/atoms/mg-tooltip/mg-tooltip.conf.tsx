export enum Guard {
  FOCUS = 'focus',
  HOVER_TOOLTIP_ELEMENT = 'hoverTooltipGuard',
  HOVER_TOOLTIPED_ELEMENT = 'hoverTooltipedGuard',
}
