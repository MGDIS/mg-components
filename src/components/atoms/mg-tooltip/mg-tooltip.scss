mg-tooltip {
  display: contents;
}

.mg-tooltip {
  // Display
  display: none;
  padding: 0.5rem;
  z-index: 9999;
  // Decoration
  border-radius: 0.3rem;
  background-color: hsl(var(--mg-tooltip-background-color));
  // Font
  color: hsl(var(--mg-tooltip-font-color));

  &[data-show] {
    display: block;
  }

  .mg-tooltip__arrow,
  .mg-tooltip__arrow:before {
    // Display
    position: absolute;
    width: 0.8rem;
    height: 0.8rem;
    z-index: 9999;
    // Decoration
    background: inherit;
  }

  .mg-tooltip__arrow {
    visibility: hidden;
  }

  .mg-tooltip__arrow:before {
    visibility: visible;
    content: '';
    transform: rotate(45deg);
  }

  &[data-popper-placement^='top'] > .mg-tooltip__arrow {
    bottom: -4px;
  }

  &[data-popper-placement^='bottom'] > .mg-tooltip__arrow {
    top: -4px;
  }

  &[data-popper-placement^='left'] > .mg-tooltip__arrow {
    right: -4px;
  }

  &[data-popper-placement^='right'] > .mg-tooltip__arrow {
    left: -4px;
  }
}
