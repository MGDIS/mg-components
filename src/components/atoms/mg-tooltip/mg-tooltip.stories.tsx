import { h } from "@stencil/core";

export default {
  component: 'mg-tooltip',
  title: 'Atoms/mg-tooltip',
  parameters: {
    layout: 'centered',
  },
};

const Template = args => <mg-tooltip {...args}><mg-icon icon="info"></mg-icon></mg-tooltip>;

export const MgTooltip = Template.bind({});

MgTooltip.args = {
  identifier: 'identifier',
  message: 'This is a tooltip message',
  placement: 'bottom',
};
