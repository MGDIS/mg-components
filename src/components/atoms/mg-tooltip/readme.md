## Usage

Tooltips are messages that provide additional information about an element.
They are presented as a message that appears when an element is hovered over or when the keyboard is focused.
Tooltips are called "custom" when they are not built on the basis of the standard HTML code provided for these elements by the specification: the title attribute.
Our "custom tooltip" component is called "tooltip" here.

## Behavior

The tooltip must be displayed when the element that allows its display :

Is hovered over by the mouse.
Takes the keyboard focus.

The tooltip must be hidden when the element that allows its display :

Is no longer hovered over by the mouse.
Loses the keyboard focus.
Pressing the Escape key must hide the tooltip.

The tooltip must remain displayed when its content is hovered over by the mouse.

The tooltip can be placed above, to the right, below or to the left of the element it describes.

By default, the message is displayed at the bottom of the element.

If the element is located at the edge of the screen, the tooltip is shifted to be always visible.

## Specs

### Shapes

![](./mg-tooltip/doc/img/mg-tooltip-shape.png)

### Fonts

![](./mg-tooltip/doc/img/mg-tooltip-font.png)

### Spacing

![](./mg-tooltip/doc/img/mg-tooltip-spaces.png)

### Alignments

![](./mg-tooltip/doc/img/mg-tooltip-alignments-text.png)

### Positioning

Triangle is always centered on the call component

![](./mg-tooltip/doc/img/mg-tooltip-positioning.png)
![](./mg-tooltip/doc/img/mg-tooltip-positioning2.png)

### Colors

![](./mg-tooltip/doc/img/mg-tooltip-colors.png)

### Sizes

![](./mg-tooltip/doc/img/mg-tooltip-sizing.png)

<!-- Auto Generated Below -->


## Properties

| Property               | Attribute    | Description                                                                      | Type                                                                                                                                                                                                         | Default                  |
| ---------------------- | ------------ | -------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------------------------ |
| `disabled`             | `disabled`   | Disable tooltip                                                                  | `boolean`                                                                                                                                                                                                    | `false`                  |
| `display`              | `display`    | Display tooltip                                                                  | `boolean`                                                                                                                                                                                                    | `false`                  |
| `identifier`           | `identifier` | Sets an `id` attribute. Needed by the input for accessibility `aria-decribedby`. | `string`                                                                                                                                                                                                     | `createID('mg-tooltip')` |
| `message` _(required)_ | `message`    | Displayed message in the tooltip                                                 | `string`                                                                                                                                                                                                     | `undefined`              |
| `placement`            | `placement`  | Tooltip placement                                                                | `"auto" \| "auto-end" \| "auto-start" \| "bottom" \| "bottom-end" \| "bottom-start" \| "left" \| "left-end" \| "left-start" \| "right" \| "right-end" \| "right-start" \| "top" \| "top-end" \| "top-start"` | `'bottom'`               |


## Dependencies

### Used by

 - [mg-input-checkbox](../../molecules/inputs/mg-input-checkbox)
 - [mg-input-date](../../molecules/inputs/mg-input-date)
 - [mg-input-numeric](../../molecules/inputs/mg-input-numeric)
 - [mg-input-password](../../molecules/inputs/mg-input-password)
 - [mg-input-radio](../../molecules/inputs/mg-input-radio)
 - [mg-input-select](../../molecules/inputs/mg-input-select)
 - [mg-input-text](../../molecules/inputs/mg-input-text)
 - [mg-input-textarea](../../molecules/inputs/mg-input-textarea)
 - [mg-input-toggle](../../molecules/inputs/mg-input-toggle)

### Graph
```mermaid
graph TD;
  mg-input-checkbox --> mg-tooltip
  mg-input-date --> mg-tooltip
  mg-input-numeric --> mg-tooltip
  mg-input-password --> mg-tooltip
  mg-input-radio --> mg-tooltip
  mg-input-select --> mg-tooltip
  mg-input-text --> mg-tooltip
  mg-input-textarea --> mg-tooltip
  mg-input-toggle --> mg-tooltip
  style mg-tooltip fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
