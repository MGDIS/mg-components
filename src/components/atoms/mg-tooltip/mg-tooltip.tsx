import { Component, Element, h, Host, Prop, Watch } from '@stencil/core';
import { createID } from '../../../utils/components.utils';
import { Instance as PopperInstance, createPopper, Placement } from '@popperjs/core';
import { Guard } from './mg-tooltip.conf';

@Component({
  tag: 'mg-tooltip',
  styleUrl: 'mg-tooltip.scss',
  shadow: true,
})
export class MgTooltip {
  /************
   * Internal *
   ************/

  private popper: PopperInstance;
  private tooltip: HTMLElement;

  // tooltip actions guards
  private guard: Guard;

  /**************
   * Decorators *
   **************/

  /**
   * Get component DOM element
   */
  @Element() element: HTMLMgTooltipElement;

  /**
   * Sets an `id` attribute.
   * Needed by the input for accessibility `aria-decribedby`.
   */
  @Prop() identifier: string = createID('mg-tooltip');

  /**
   * Displayed message in the tooltip
   */
  @Prop() message!: string;
  @Watch('message')
  validateMessage(newValue: string): void {
    if (typeof newValue !== 'string' || newValue.trim() === '') {
      throw new Error('<mg-tooltip> prop "message" is required.');
    }
  }

  /**
   * Tooltip placement
   */
  @Prop() placement: Placement = 'bottom';

  /**
   * Display tooltip
   */
  @Prop({ mutable: true }) display = false;
  @Watch('display')
  handleDisplay(newValue: boolean): void {
    if (newValue) {
      this.show();
    } else {
      this.hide();
    }
  }

  /**
   * Disable tooltip
   */
  @Prop({ mutable: true }) disabled = false;

  /**
   * Show tooltip
   *
   * @returns {void}
   */
  private show = (): void => {
    // Make the tooltip visible
    this.tooltip.setAttribute('data-show', '');
    // Enable the event listeners
    this.popper.setOptions(options => ({
      ...options,
      modifiers: [...options.modifiers, { name: 'eventListeners', enabled: true }],
    }));
    // Update its position
    this.popper.update();
  };

  /**
   * Hide tooltip
   *
   * @returns {void}
   */
  private hide = (): void => {
    // Hide the tooltip
    this.tooltip.removeAttribute('data-show');
    // Disable the event listeners
    this.popper.setOptions(options => ({
      ...options,
      modifiers: [...options.modifiers, { name: 'eventListeners', enabled: false }],
    }));
  };

  /**
   * Method to set display prop
   *
   * @param {boolean} newValue display prop new value
   * @param {boolean} condition additionnal condition to apply display prop newValue
   * @returns {void}
   */
  private setDisplay = (newValue: boolean, condition = true): void => {
    if (!this.disabled && condition) this.display = newValue;
  };

  /**
   * Action for tooltip element and tooltiped element mouse listener
   *
   * @param {Guard} elementGuard tooltip element guard
   * @param {boolean} isMouseenter mouseenter validation
   * @param {Guard} conditionalGuard guard condition
   * @returns {void}
   */
  private tooltipMouseListenerAction = (elementGuard: Guard, isMouseenter: boolean, conditionalGuard: Guard): void => {
    // we mutate elementGuard
    if (this.guard !== Guard.FOCUS) {
      this.guard = elementGuard;
      if (!isMouseenter) {
        setTimeout(() => {
          this.setDisplay(isMouseenter, this.guard !== conditionalGuard);
          this.resetGuard();
        }, 100);
      } else if (isMouseenter && elementGuard === Guard.HOVER_TOOLTIPED_ELEMENT) {
        this.setDisplay(isMouseenter);
      }
    }
  };

  /**
   * Method to reset guard value
   *
   * @returns {void}
   */
  private resetGuard = (): void => {
    this.guard = undefined;
  };

  /*************
   * Lifecycle *
   *************/

  /**
   * Get slotted element
   * Check if it already contain an interactive element, if not we need to add a tabIndex attribute
   * We need to attach the focused element to the tooltip (aria-describedby)
   *
   * @returns {void}
   */
  componentDidLoad(): void {
    // Get tooltip element
    this.tooltip = this.element.shadowRoot.querySelector(`#${this.identifier}`);

    // get slotted element
    const slotElement = this.element.firstElementChild as HTMLElement;
    // Get interactive element
    const interactiveElements = ['a', 'button', 'input', 'textarea', 'select']; //! Might needs updates
    const interactiveElement = this.element.shadowRoot.querySelector(interactiveElements.join(',')) || slotElement.shadowRoot?.querySelector(interactiveElements.join(','));

    // define selected element to become tooltip selector
    const tooltipedElement = interactiveElement || slotElement;

    // Add tabindex to slotted element if we can't find any interactive element
    if (interactiveElement === null || interactiveElement === undefined) slotElement.tabIndex = 0;

    // Set aria-describedby
    const ariaDescribedby = slotElement.getAttribute('aria-describedby');
    if (ariaDescribedby === null) {
      tooltipedElement.setAttribute('aria-describedby', this.identifier);
    } else {
      slotElement.setAttribute('aria-describedby', `${ariaDescribedby} ${this.identifier}`);
    }

    // Create popperjs tooltip
    this.popper = createPopper(tooltipedElement, this.tooltip, {
      placement: this.placement,
      strategy: 'fixed',
      modifiers: [
        {
          name: 'offset',
          options: {
            offset: [0, 8],
          },
        },
      ],
    });

    // Manage tooltipedElement focus/blur events
    tooltipedElement.addEventListener('focus', () => {
      this.guard = Guard.FOCUS;
      this.setDisplay(true);
    });

    ['blur', 'keydown'].forEach(event => {
      tooltipedElement.addEventListener(event, (e: UIEvent & KeyboardEvent) => {
        // we continue to process ONLY for KeyboardEvents 'Escape'
        if (e.type === 'keydown' && e.code !== 'Escape') {
          return;
        }
        this.resetGuard();
        this.setDisplay(false);
      });
    });

    // manage tooltipElement & tooltipedElement mouseenter/mouseleave events
    ['mouseenter', 'mouseleave'].forEach(event => {
      const isMouseenter = event === 'mouseenter';
      [
        { element: this.tooltip, action: () => this.tooltipMouseListenerAction(Guard.HOVER_TOOLTIP_ELEMENT, isMouseenter, Guard.HOVER_TOOLTIPED_ELEMENT) },
        { element: tooltipedElement, action: () => this.tooltipMouseListenerAction(Guard.HOVER_TOOLTIPED_ELEMENT, isMouseenter, Guard.HOVER_TOOLTIP_ELEMENT) },
      ].forEach(({ element, action }) => {
        element.addEventListener(event, () => {
          action();
        });
      });
    });

    this.handleDisplay(this.display);
    this.validateMessage(this.message);
  }

  /**
   * Render
   *
   * @returns {HTMLElement} HTML Element
   */
  render(): HTMLElement {
    return (
      <Host>
        <slot></slot>
        <div role="tooltip" id={this.identifier} class="mg-tooltip" innerHTML={this.message}>
          <div class="mg-tooltip__arrow" data-popper-arrow></div>
        </div>
      </Host>
    );
  }
}
