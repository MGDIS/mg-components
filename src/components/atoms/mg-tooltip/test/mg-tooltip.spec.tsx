import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgTooltip } from '../mg-tooltip';
import { MgButton } from '../../mg-button/mg-button';
import { MgIcon } from '../../mg-icon/mg-icon';

// fix popper console.error in test
// it is generated in @popperjs/core/dist/cjs/popper.js l.1859
// this is due to internal function isHTMLElement(), so we can not mock it directly.
// this function check if test DOM element mockHTMLElement instance is 'instanceof HTMLElement'
// so we only override the console.error side effect for this error
const errorFunction = console.error;
const mock = jest.spyOn(console, 'error')
mock.mockImplementation((error) => {
  const compareWith = 'Popper: "arrow" element must be an HTMLElement (not an SVGElement). To use an SVG arrow, wrap it in an HTMLElement that will be used as the arrow.';
  if (error !== compareWith) {
    errorFunction(error)
  }
})

const getPage = (args, element) => newSpecPage({
  components: [MgTooltip, MgButton, MgIcon],
  template: () => (<mg-tooltip {...args}>{element}</mg-tooltip>),

});

describe('mg-tooltip', () => {
  test.each([
    <span>span</span>,
    <button aria-describedby="blu">button</button>,
    <mg-icon icon="success"></mg-icon>,
    <mg-button identifier="identifier">mg-button</mg-button>
  ])('Should render with element', async (element) => {
    const args = { identifier:"identifier", message: 'My tooltip message'};
    const {root} = await getPage(args, element);
    expect(root).toMatchSnapshot();
  });

  describe.each([
    {eventIn:'mouseenter', eventOut:'mouseleave'},
    {eventIn:'focus', eventOut:'blur'},
  ])('Should manage display on events enter with %s, leave with %s', ({eventIn, eventOut}) => {
    test.each([
      <span>span</span>,
      <button aria-describedby="blu">button</button>,
      <mg-icon icon="success"></mg-icon>,
      <mg-button identifier="identifier">mg-button</mg-button>
    ])('element', async(element) => {
      const args = { identifier:"identifier", message: 'blu'};
      const page = await getPage(args, element);
      const mgTooltip = page.doc.querySelector('mg-tooltip');
      const linkedTooltipElement = mgTooltip.querySelector(`[aria-describedby*='${args.identifier}']`);
      const tooltip = mgTooltip.querySelector(`#${args.identifier}`);

      linkedTooltipElement.dispatchEvent(new CustomEvent(eventIn, { bubbles: true }));
      await page.waitForChanges();

      expect(page.root).toMatchSnapshot();
      expect(tooltip).toHaveAttribute('data-show');

      linkedTooltipElement.dispatchEvent(new CustomEvent(eventOut, { bubbles: true }));
      await page.waitForChanges();

      expect(tooltip).not.toHaveAttribute('data-show');
    });
  });

  test.each([true, false])('Should toogle tooltip from prop display, case display %s', async (display) => {
    const args = { identifier:"identifier", message: 'batman', display};
    const element = <span>batman</span>;
    const page = await getPage(args, element);

    const mgTooltip = page.doc.querySelector('mg-tooltip');
    const tooltip = mgTooltip.querySelector(`#${args.identifier}`);

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
    if(display) {
      expect(tooltip).toHaveAttribute('data-show');
    } else {
      expect(tooltip).not.toHaveAttribute('data-show');
    }

    page.rootInstance.display = !display;
    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
    if(display) {
      expect(tooltip).not.toHaveAttribute('data-show');
    } else {
      expect(tooltip).toHaveAttribute('data-show');
    }
  });

  describe('method hide', () => {
    test.each(['Tab', 'Space', 'Enter' , 'i'])('should prevent keyboardEvent, case not "Escape" code', async (key) => {
      const args = { identifier:"identifier", message: 'batman'};
      const element = <span id="batman">batman</span>;
      const page = await getPage(args, element);

      const mgTooltip = page.doc.querySelector('mg-tooltip');
      const tooltip = mgTooltip.querySelector(`#${args.identifier}`);
      const i = page.doc.querySelector('#batman')

      i.dispatchEvent(new CustomEvent('focus', { bubbles: true }));
      await page.waitForChanges();

      expect(tooltip).toHaveAttribute('data-show');

      i.dispatchEvent(new KeyboardEvent('keydown', {'code': key}));
      await page.waitForChanges();

      expect(tooltip).toHaveAttribute('data-show');
    })
  })
});

