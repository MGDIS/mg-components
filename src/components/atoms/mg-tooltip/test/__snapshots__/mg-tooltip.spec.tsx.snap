// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-tooltip Should manage display on events enter with { eventIn: 'focus', eventOut: 'blur' }, leave with %s element 1`] = `
<mg-tooltip>
  <!---->
  <span aria-describedby="identifier" tabindex="0">
    span
  </span>
  <div class="mg-tooltip" data-show id="identifier" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0;">
    blu
    <div class="mg-tooltip__arrow" data-popper-arrow style="position: absolute;"></div>
  </div>
</mg-tooltip>
`;

exports[`mg-tooltip Should manage display on events enter with { eventIn: 'focus', eventOut: 'blur' }, leave with %s element 2`] = `
<mg-tooltip>
  <!---->
  <button aria-describedby="blu identifier">
    button
  </button>
  <div class="mg-tooltip" data-show id="identifier" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0;">
    blu
    <div class="mg-tooltip__arrow" data-popper-arrow style="position: absolute;"></div>
  </div>
</mg-tooltip>
`;

exports[`mg-tooltip Should manage display on events enter with { eventIn: 'focus', eventOut: 'blur' }, leave with %s element 3`] = `
<mg-tooltip>
  <!---->
  <mg-icon aria-describedby="identifier" tabindex="0">
    <mock:shadow-root>
      <svg aria-hidden="true" class="mg-icon mg-icon--check-circle mg-icon--size-regular" focusable="false" viewBox="0 0 16 16">
        <path d="M15.33 8A7.33 7.33 0 1 1 8 .67 7.33 7.33 0 0 1 15.33 8Zm-8.18 3.88 5.44-5.44a.46.46 0 0 0 0-.65l-.67-.67a.48.48 0 0 0-.67 0L6.82 9.54 4.75 7.47a.46.46 0 0 0-.65 0l-.67.67a.46.46 0 0 0 0 .65l3.07 3.07a.46.46 0 0 0 .65 0Z" fill="currentColor"></path>
      </svg>
    </mock:shadow-root>
  </mg-icon>
  <div class="mg-tooltip" data-show id="identifier" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0;">
    blu
    <div class="mg-tooltip__arrow" data-popper-arrow style="position: absolute;"></div>
  </div>
</mg-tooltip>
`;

exports[`mg-tooltip Should manage display on events enter with { eventIn: 'focus', eventOut: 'blur' }, leave with %s element 4`] = `
<mg-tooltip>
  <!---->
  <mg-button>
    <!---->
    <button aria-describedby="identifier" class="mg-button mg-button--primary" id="identifier-mg-button">
      <div class="mg-button__content">
        mg-button
      </div>
    </button>
  </mg-button>
  <div class="mg-tooltip" data-show id="identifier" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0;">
    blu
    <div class="mg-tooltip__arrow" data-popper-arrow style="position: absolute;"></div>
  </div>
</mg-tooltip>
`;

exports[`mg-tooltip Should manage display on events enter with { eventIn: 'mouseenter', eventOut: 'mouseleave' }, leave with %s element 1`] = `
<mg-tooltip>
  <!---->
  <span aria-describedby="identifier" tabindex="0">
    span
  </span>
  <div class="mg-tooltip" data-show id="identifier" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0;">
    blu
    <div class="mg-tooltip__arrow" data-popper-arrow style="position: absolute;"></div>
  </div>
</mg-tooltip>
`;

exports[`mg-tooltip Should manage display on events enter with { eventIn: 'mouseenter', eventOut: 'mouseleave' }, leave with %s element 2`] = `
<mg-tooltip>
  <!---->
  <button aria-describedby="blu identifier">
    button
  </button>
  <div class="mg-tooltip" data-show id="identifier" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0;">
    blu
    <div class="mg-tooltip__arrow" data-popper-arrow style="position: absolute;"></div>
  </div>
</mg-tooltip>
`;

exports[`mg-tooltip Should manage display on events enter with { eventIn: 'mouseenter', eventOut: 'mouseleave' }, leave with %s element 3`] = `
<mg-tooltip>
  <!---->
  <mg-icon aria-describedby="identifier" tabindex="0">
    <mock:shadow-root>
      <svg aria-hidden="true" class="mg-icon mg-icon--check-circle mg-icon--size-regular" focusable="false" viewBox="0 0 16 16">
        <path d="M15.33 8A7.33 7.33 0 1 1 8 .67 7.33 7.33 0 0 1 15.33 8Zm-8.18 3.88 5.44-5.44a.46.46 0 0 0 0-.65l-.67-.67a.48.48 0 0 0-.67 0L6.82 9.54 4.75 7.47a.46.46 0 0 0-.65 0l-.67.67a.46.46 0 0 0 0 .65l3.07 3.07a.46.46 0 0 0 .65 0Z" fill="currentColor"></path>
      </svg>
    </mock:shadow-root>
  </mg-icon>
  <div class="mg-tooltip" data-show id="identifier" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0;">
    blu
    <div class="mg-tooltip__arrow" data-popper-arrow style="position: absolute;"></div>
  </div>
</mg-tooltip>
`;

exports[`mg-tooltip Should manage display on events enter with { eventIn: 'mouseenter', eventOut: 'mouseleave' }, leave with %s element 4`] = `
<mg-tooltip>
  <!---->
  <mg-button>
    <!---->
    <button aria-describedby="identifier" class="mg-button mg-button--primary" id="identifier-mg-button">
      <div class="mg-button__content">
        mg-button
      </div>
    </button>
  </mg-button>
  <div class="mg-tooltip" data-show id="identifier" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0;">
    blu
    <div class="mg-tooltip__arrow" data-popper-arrow style="position: absolute;"></div>
  </div>
</mg-tooltip>
`;

exports[`mg-tooltip Should render with element 1`] = `
<mg-tooltip>
  <!---->
  <span aria-describedby="identifier" tabindex="0">
    span
  </span>
  <div class="mg-tooltip" id="identifier" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0;">
    My tooltip message
    <div class="mg-tooltip__arrow" data-popper-arrow style="position: absolute;"></div>
  </div>
</mg-tooltip>
`;

exports[`mg-tooltip Should render with element 2`] = `
<mg-tooltip>
  <!---->
  <button aria-describedby="blu identifier">
    button
  </button>
  <div class="mg-tooltip" id="identifier" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0;">
    My tooltip message
    <div class="mg-tooltip__arrow" data-popper-arrow style="position: absolute;"></div>
  </div>
</mg-tooltip>
`;

exports[`mg-tooltip Should render with element 3`] = `
<mg-tooltip>
  <!---->
  <mg-icon aria-describedby="identifier" tabindex="0">
    <mock:shadow-root>
      <svg aria-hidden="true" class="mg-icon mg-icon--check-circle mg-icon--size-regular" focusable="false" viewBox="0 0 16 16">
        <path d="M15.33 8A7.33 7.33 0 1 1 8 .67 7.33 7.33 0 0 1 15.33 8Zm-8.18 3.88 5.44-5.44a.46.46 0 0 0 0-.65l-.67-.67a.48.48 0 0 0-.67 0L6.82 9.54 4.75 7.47a.46.46 0 0 0-.65 0l-.67.67a.46.46 0 0 0 0 .65l3.07 3.07a.46.46 0 0 0 .65 0Z" fill="currentColor"></path>
      </svg>
    </mock:shadow-root>
  </mg-icon>
  <div class="mg-tooltip" id="identifier" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0;">
    My tooltip message
    <div class="mg-tooltip__arrow" data-popper-arrow style="position: absolute;"></div>
  </div>
</mg-tooltip>
`;

exports[`mg-tooltip Should render with element 4`] = `
<mg-tooltip>
  <!---->
  <mg-button>
    <!---->
    <button aria-describedby="identifier" class="mg-button mg-button--primary" id="identifier" style="position: absolute; left: 0; top: 0; margin: 0;">
      <div class="mg-button__content">
        mg-button
      </div>
    </button>
  </mg-button>
  <div class="mg-tooltip" id="identifier" role="tooltip">
    My tooltip message
    <div class="mg-tooltip__arrow" data-popper-arrow></div>
  </div>
</mg-tooltip>
`;

exports[`mg-tooltip Should toogle tooltip from prop display, case display false 1`] = `
<mg-tooltip>
  <!---->
  <span aria-describedby="identifier" tabindex="0">
    batman
  </span>
  <div class="mg-tooltip" id="identifier" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0;">
    batman
    <div class="mg-tooltip__arrow" data-popper-arrow style="position: absolute;"></div>
  </div>
</mg-tooltip>
`;

exports[`mg-tooltip Should toogle tooltip from prop display, case display false 2`] = `
<mg-tooltip>
  <!---->
  <span aria-describedby="identifier" tabindex="0">
    batman
  </span>
  <div class="mg-tooltip" data-show id="identifier" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0;">
    batman
    <div class="mg-tooltip__arrow" data-popper-arrow style="position: absolute;"></div>
  </div>
</mg-tooltip>
`;

exports[`mg-tooltip Should toogle tooltip from prop display, case display true 1`] = `
<mg-tooltip>
  <!---->
  <span aria-describedby="identifier" tabindex="0">
    batman
  </span>
  <div class="mg-tooltip" data-show id="identifier" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0;">
    batman
    <div class="mg-tooltip__arrow" data-popper-arrow style="position: absolute;"></div>
  </div>
</mg-tooltip>
`;

exports[`mg-tooltip Should toogle tooltip from prop display, case display true 2`] = `
<mg-tooltip>
  <!---->
  <span aria-describedby="identifier" tabindex="0">
    batman
  </span>
  <div class="mg-tooltip" id="identifier" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0;">
    batman
    <div class="mg-tooltip__arrow" data-popper-arrow style="position: absolute;"></div>
  </div>
</mg-tooltip>
`;
